Add an analyzer rule against unused using directives

penev92 opened the pull request:
What started as simple "spring" cleaning turned out bigger than expected, so splitting it into its own PR.
Unfortunately none of the options I tried with the suggested analyzer rules `IDE0005` and `CS8019` could make `make check` fail in any way, but I still think having the rule is an improvement. Arguably it could be even made to show as an error inside an IDE (both VS and VSCode do it), but it still won't stop the build.

penev92 committed:
Removed unused using directives

pchote commented:
Did you try defining the full
```xml
<Rules AnalyzerId="Microsoft.CodeAnalysis.CSharp.Features" RuleNamespace="Microsoft.CodeAnalysis.CSharp.Features">
   <Rule Id="IDE0005" Action="Warning" />
</Rules>
```
?

IDE0005 isn't part of the `StyleCop.Analyzers` namespace, so it seems to me that `make check` is behaving correctly and your IDE is wrong.

penev92 commented:
Tested both namespaces (`StyleCop.Analyzers` and `Microsoft.CodeAnalysis.CSharp.Features`) with both rules (`IDE0005` and `CS8019`) and no combination of any two or more behaves any different to the other combinations - both IDEs still report a warning/error, but still compile successfully and `dotnet build` still reports 0 warnings/errors and compiles successfully.

RoosterDragon commented:
So it looks like IDE0005 *specifically* is borked. [IDE0005 docs](https://docs.microsoft.com/en-gb/dotnet/fundamentals/code-analysis/style-rules/ide0005) have a note which states `To enable this rule on build, you need to enable XML documentation comments for the project. See this issue for more details.` Which links to https://github.com/dotnet/roslyn/issues/41640

If you follow the advice, you can indeed get it to show up at build time. Add this to the `csproj`
```xml
<PropertyGroup>
    <GenerateDocumentationFile>true</GenerateDocumentationFile>
</PropertyGroup>
```

You'll get some additional spam but it does actually work!

Therefore your changes are fine, it's just a bug in the infrastructure. I wouldn't bother with the workaround since it adds other spam - I think adding the rules will do for now and at some point in future I guess it will magically work in the build again. Other IDEXXX rules are not affected and appear to work without any workarounds.

That said I agree with pchote it would be best to put it into the correct analyser group. Apparently you can choose between `Microsoft.CodeAnalysis.CSharp.Features` and `Microsoft.CodeAnalysis.CSharp.CodeStyle` - they seem to have the same rules and no I don't know what the difference is.

```xml
<Rules AnalyzerId="Microsoft.CodeAnalysis.CSharp.CodeStyle" RuleNamespace="Microsoft.CodeAnalysis.CSharp.CodeStyle">
    <Rule Id="IDE0005" Action="Warning" />
</Rules>
```

Changes looks fine otherwise.

penev92 committed:
Added a rule for unused usings to OpenRA.ruleset

Unfortunately due to bugs in the analyzers or something else, the IDE0005 doesn't work as expected. The "officially suggested" workaround is to enable XML documentation generation to trigger IDE0005 during compiling. Then we need to add three more rules to silence the warnings that come from the XML documentation generation. We also need to enable code style enforcing on build for all of this to work.
Known issue is that all of this produces a bunch (tens to hundreds) of obscure analyzer warnings on older versions of Visual Studio, but those seem to not be causing issues.

penev92 commented:
Thanks to @RoosterDragon we now know we need a combination of `EnforceCodeStyleInBuild` and `GenerateDocumentationFile` in the project file/`Directory.Build.props` as well as the rule itself, and then a bunch more rules to silence the warnings that come from generating the XML documentation.
I moved rule `IDE0005` to the namespace mentioned above.
It all works as expected now with one known issue - `EnforceCodeStyleInBuild` produces a bunch of warnings in older versions of Visual Studio. I tested an old and a new version of VS, tested in VSCode and tested the Windows `make` script and everything seems to work as it should.

pchote commented:
We could avoid the VS problems by specifying `-p:EnforceCodeStyleInBuild=true -p:GenerateDocumentationFile=true` in the msbuild args in the check rule only, like we did for #19671.

I'm pretty sure the CS rules also don't live in StyleCop's namespace?

RoosterDragon commented:
Can confirm the above would work. We still get the warning in the IDE, but `make check` will also be able to surface it. We would need to maintain the CSXXXX overrides though, or else `make check` will spit those out too.

pchote commented:
Yeah, that seems like the best compromise to me - they just need to go in the right namespace 😁

RoosterDragon commented:
This is the namespace you'll want for compiler rules.

```xml
<Rules AnalyzerId="Microsoft.CodeAnalysis.CSharp" RuleNamespace="Microsoft.CodeAnalysis.CSharp">
    <Rule Id="CS1570" Action="None" />
    <Rule Id="CS1573" Action="None" />
    <Rule Id="CS1591" Action="None" />
</Rules>
```

penev92 commented:
Added a separate "group" for the CSxxxx (XML documentation generation) rules with the proper namespace as provided above.
Also added explanations to the new rules and a bunch of empty lines for improved readability.
Also made the XML generation (and code style enforcement) only run when compiling in Debug.

penev92 commented:
*Also made the XML generation (and code style enforcement) only run for the `make check` command and added a comment there explaining the weird new parameters.

pchote commented on Makefile:
```diff
@@ -99,9 +99,11 @@ check:
 	@echo "Compiling in Debug mode..."
 ifeq ($(RUNTIME), mono)
-	@$(MSBUILD) -t:build -restore -p:Configuration=Debug -warnaserror -p:TargetPlatform=$(TARGETPLATFORM) -p:Mono=true
+	# Enabling EnforceCodeStyleInBuild and GenerateDocumentationFile as a workaround for some code style rules (in particular IDE0005) being bugged and not reporting warnings/errors otherwise.
```
Please either remove the tab indent from these these two comments, or prefix them with `@` so that they aren't literally printed to the terminal (`make` is special).

penev92 replied:
Done

pchote reviewed (commented):
LGTM otherwise, confirmed to work as expected on macOS.

pchote reviewed (approved):
LGTM now (but i'm trusting that you've tested it works correctly on Windows).
